src/docfx/build/manifest/ManifestBuilder.cs

```diff
@@ -48,33 +38,39 @@ public bool TryAdd(Document file, FileManifest manifest)
                 return false;
             }
 
+            _filesBySiteUrl.GetOrAdd(manifest.SiteUrl, _ => new ConcurrentBag<(Document doc, List<string> monikers)>()).Add((file, monikers));
+
             return true;
         }
 
         public (FileManifest[], List<Document>) Build(Context context)
         {
             // Handle publish conflicts
-            foreach (var (siteUrl, conflict) in _siteUrlConflicts)
+            foreach (var (siteUrl, files) in _filesBySiteUrl)
             {
-                var conflictingFiles = new HashSet<Document>();
-
-                foreach (var conflictingFile in conflict)
+                var hasConflict = false;
+                var uniqueFiles = files.GroupBy(file => file.doc).ToList();
+                for (var i = 0; i < uniqueFiles.Count; i++)
                 {
-                    conflictingFiles.Add(conflictingFile);
+                    var firstMonikers = uniqueFiles[i].FirstOrDefault().monikers;
+                    if (uniqueFiles.Skip(i + 1).Any(file => CheckMonikerConflict(firstMonikers, file.FirstOrDefault().monikers)))
+                    {
+                        hasConflict = true;
+                        break;
+                    }
```

Comment:
try to finish the Conflict check in one loop, it's shouls be possible if you don't need know detailed conflict info